File: dotnet/example_code/DynamoDB/SampleDataLoad.cs
// snippet-sourcedescription:[SampleDataLoad.cs demonstrates how to ]
// snippet-service:[dynamodb]
// snippet-keyword:[dotNET]
// snippet-keyword:[Amazon DynamoDB]
// snippet-keyword:[Code Sample]
// snippet-keyword:[ ]
// snippet-sourcetype:[full-example]
// snippet-sourcedate:[ ]
// snippet-sourceauthor:[AWS]
// snippet-start:[dynamodb.dotNET.CodeExample.SampleDataLoad] 

/**
 * Copyright 2010-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * This file is licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License. A copy of
 * the License is located at
 *
 * http://aws.amazon.com/apache2.0/
 *
 * This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
*/
using System;
using System.Collections.Generic;
using Amazon.DynamoDBv2;
using Amazon.DynamoDBv2.DocumentModel;
using Amazon.Runtime;
using Amazon.SecurityToken;

namespace com.amazonaws.codesamples
{
    class SampleDataLoad
    {
        private static AmazonDynamoDBClient client = new AmazonDynamoDBClient();

        static void Main(string[] args)
        {
            try
            {
                // Load data (using the .NET document API)
                LoadSampleProducts();
                LoadSampleForums();
                LoadSampleThreads();
                LoadSampleReplies();

                Console.WriteLine("Data loaded... To continue, press Enter");
                Console.ReadLine();
            }
            catch (AmazonDynamoDBException e) { Console.WriteLine(e.Message); }
            catch (AmazonServiceException e) { Console.WriteLine(e.Message); }
            catch (Exception e) { Console.WriteLine(e.Message); }
        }

        private static void LoadSampleProducts()
        {
            Table productCatalogTable = Table.LoadTable(client, "ProductCatalog");
            // ********** Add Books *********************
            var book1 = new Document();
            book1["Id"] = 101;
            book1["Title"] = "Book 101 Title";
            book1["ISBN"] = "111-1111111111";
            book1["Authors"] = new List<string> { "Author 1" };
            book1["Price"] = -2; // *** Intentional value. Later used to illustrate scan.
            book1["Dimensions"] = "8.5 x 11.0 x 0.5";
            book1["PageCount"] = 500;
            book1["InPublication"] = true;
            book1["ProductCategory"] = "Book";
            productCatalogTable.PutItem(book1);

            var book2 = new Document();

            book2["Id"] = 102;
            book2["Title"] = "Book 102 Title";
            book2["ISBN"] = "222-2222222222";
            book2["Authors"] = new List<string> { "Author 1", "Author 2" }; ;
            book2["Price"] = 20;
            book2["Dimensions"] = "8.5 x 11.0 x 0.8";
            book2["PageCount"] = 600;
            book2["InPublication"] = true;
            book2["ProductCategory"] = "Book";
            productCatalogTable.PutItem(book2);

            var book3 = new Document();
            book3["Id"] = 103;
            book3["Title"] = "Book 103 Title";
            book3["ISBN"] = "333-3333333333";
            book3["Authors"] = new List<string> { "Author 1", "Author2", "Author 3" }; ;
            book3["Price"] = 2000;
            book3["Dimensions"] = "8.5 x 11.0 x 1.5";
            book3["PageCount"] = 700;
            book3["InPublication"] = false;
            book3["ProductCategory"] = "Book";
            productCatalogTable.PutItem(book3);

            // ************ Add bikes. *******************
            var bicycle1 = new Document();
            bicycle1["Id"] = 201;
            bicycle1["Title"] = "18-Bike 201"; // size, followed by some title.
            bicycle1["Description"] = "201 description";
            bicycle1["BicycleType"] = "Road";
            bicycle1["Brand"] = "Brand-Company A"; // Trek, Specialized.
            bicycle1["Price"] = 100;
            bicycle1["Color"] = new List<string> { "Red", "Black" };
            bicycle1["ProductCategory"] = "Bike";
            productCatalogTable.PutItem(bicycle1);

            var bicycle2 = new Document();
            bicycle2["Id"] = 202;
            bicycle2["Title"] = "21-Bike 202Brand-Company A";
            bicycle2["Description"] = "202 description";
            bicycle2["BicycleType"] = "Road";
            bicycle2["Brand"] = "";
            bicycle2["Price"] = 200;
            bicycle2["Color"] = new List<string> { "Green", "Black" };
            bicycle2["ProductCategory"] = "Bicycle";
            productCatalogTable.PutItem(bicycle2);

            var bicycle3 = new Document();
            bicycle3["Id"] = 203;
            bicycle3["Title"] = "19-Bike 203";
            bicycle3["Description"] = "203 description";
            bicycle3["BicycleType"] = "Road";
            bicycle3["Brand"] = "Brand-Company B";
            bicycle3["Price"] = 300;
            bicycle3["Color"] = new List<string> { "Red", "Green", "Black" };
            bicycle3["ProductCategory"] = "Bike";
            productCatalogTable.PutItem(bicycle3);

            var bicycle4 = new Document();
            bicycle4["Id"] = 204;
            bicycle4["Title"] = "18-Bike 204";
            bicycle4["Description"] = "204 description";
            bicycle4["BicycleType"] = "Mountain";
            bicycle4["Brand"] = "Brand-Company B";
            bicycle4["Price"] = 400;
            bicycle4["Color"] = new List<string> { "Red" };
            bicycle4["ProductCategory"] = "Bike";
            productCatalogTable.PutItem(bicycle4);

            var bicycle5 = new Document();
            bicycle5["Id"] = 205;
            bicycle5["Title"] = "20-Title 205";
            bicycle4["Description"] = "205 description";
            bicycle5["BicycleType"] = "Hybrid";
            bicycle5["Brand"] = "Brand-Company C";
            bicycle5["Price"] = 500;
            bicycle5["Color"] = new List<string> { "Red", "Black" };
            bicycle5["ProductCategory"] = "Bike";
            productCatalogTable.PutItem(bicycle5);
        }

        private static void LoadSampleForums()
        {
            Table forumTable = Table.LoadTable(client, "Forum");

            var forum1 = new Document();
            forum1["Name"] = "Amazon DynamoDB"; // PK
            forum1["Category"] = "Amazon Web Services";
            forum1["Threads"] = 2;
            forum1["Messages"] = 4;
            forum1["Views"] = 1000;

            forumTable.PutItem(forum1);

            var forum2 = new Document();
            forum2["Name"] = "Amazon S3"; // PK
            forum2["Category"] = "Amazon Web Services";
            forum2["Threads"] = 1;

            forumTable.PutItem(forum2);
        }

        private static void LoadSampleThreads()
        {
            Table threadTable = Table.LoadTable(client, "Thread");

            // Thread 1.
            var thread1 = new Document();
            thread1["ForumName"] = "Amazon DynamoDB"; //Partition key
            thread1["Subject"] = "DynamoDB Thread 1"; //Sort key
            thread1["Message"] = "DynamoDB thread 1 message text";
            thread1["LastPostedBy"] = "User A";
            thread1["LastPostedDateTime"] = DateTime.UtcNow.Subtract(new TimeSpan(14, 0, 0, 0));
            thread1["Views"] = 0;
            thread1["Replies"] = 0;
            thread1["Answered"] = false;
            thread1["Tags"] = new List<string> { "index", "primarykey", "table" };

            threadTable.PutItem(thread1);

            // Thread 2.
            var thread2 = new Document();
            thread2["ForumName"] = "Amazon DynamoDB"; //Partition key
            thread2["Subject"] = "DynamoDB Thread 2"; //Sort key
            thread2["Message"] = "DynamoDB thread 2 message text";
            thread2["LastPostedBy"] = "User A";
            thread2["LastPostedDateTime"] = DateTime.UtcNow.Subtract(new TimeSpan(21, 0, 0, 0));
            thread2["Views"] = 0;
            thread2["Replies"] = 0;
            thread2["Answered"] = false;
            thread2["Tags"] = new List<string> { "index", "hashkey", "sortkey" };

            threadTable.PutItem(thread2);

            // Thread 3.
            var thread3 = new Document();
            thread3["ForumName"] = "Amazon S3"; // Partition key
            thread3["Subject"] = "S3 Thread 1"; // Sort key
            thread3["Message"] = "S3 thread 3 message text";
            thread3["LastPostedBy"] = "User A";
            thread3["LastPostedDateTime"] = DateTime.UtcNow.Subtract(new TimeSpan(7, 0, 0, 0));
            thread3["Views"] = 0;
            thread3["Replies"] = 0;
            thread3["Answered"] = false;
            thread3["Tags"] = new List<string> { "largeobjects", "multipart upload" };

            threadTable.PutItem(thread3);
        }

        private static void LoadSampleReplies()
        {
            Table replyTable = Table.LoadTable(client, "Reply");

            // Reply 1 - thread 1.
            var thread1Reply1 = new Document();
            thread1Reply1["Id"] = "Amazon DynamoDB#DynamoDB Thread 1"; //Partition key
            thread1Reply1["ReplyDateTime"] = DateTime.UtcNow.Subtract(new TimeSpan(21, 0, 0, 0)); //Sort key
            thread1Reply1["Message"] = "DynamoDB Thread 1 Reply 1 text";
            thread1Reply1["PostedBy"] = "User A";

            replyTable.PutItem(thread1Reply1);

            // Reply 2 - thread 1.
            var thread1reply2 = new Document();
            thread1reply2["Id"] = "Amazon DynamoDB#DynamoDB Thread 1"; //Partition key
            thread1reply2["ReplyDateTime"] = DateTime.UtcNow.Subtract(new TimeSpan(14, 0, 0, 0)); //Sort key
            thread1reply2["Message"] = "DynamoDB Thread 1 Reply 2 text";
            thread1reply2["PostedBy"] = "User B";

            replyTable.PutItem(thread1reply2);

            // Reply 3 - thread 1.
            var thread1Reply3 = new Document();
            thread1Reply3["Id"] = "Amazon DynamoDB#DynamoDB Thread 1"; //Partition key
            thread1Reply3["ReplyDateTime"] = DateTime.UtcNow.Subtract(new TimeSpan(7, 0, 0, 0)); //Sort key
            thread1Reply3["Message"] = "DynamoDB Thread 1 Reply 3 text";
            thread1Reply3["PostedBy"] = "User B";

            replyTable.PutItem(thread1Reply3);

            // Reply 1 - thread 2.
            var thread2Reply1 = new Document();
            thread2Reply1["Id"] = "Amazon DynamoDB#DynamoDB Thread 2"; //Partition key
            thread2Reply1["ReplyDateTime"] = DateTime.UtcNow.Subtract(new TimeSpan(7, 0, 0, 0)); //Sort key
            thread2Reply1["Message"] = "DynamoDB Thread 2 Reply 1 text";
            thread2Reply1["PostedBy"] = "User A";


            replyTable.PutItem(thread2Reply1);

            // Reply 2 - thread 2.
            var thread2Reply2 = new Document();
            thread2Reply2["Id"] = "Amazon DynamoDB#DynamoDB Thread 2"; //Partition key
            thread2Reply2["ReplyDateTime"] = DateTime.UtcNow.Subtract(new TimeSpan(1, 0, 0, 0)); //Sort key
            thread2Reply2["Message"] = "DynamoDB Thread 2 Reply 2 text";
            thread2Reply2["PostedBy"] = "User A";

            replyTable.PutItem(thread2Reply2);
        }
    }
}
// snippet-end:[dynamodb.dotNET.CodeExample.SampleDataLoad]

File: php/example_code/dynamodb/SampleDataLoad.php
<?php
// snippet-sourcedescription:[SampleDataLoad.php demonstrates how to ]
// snippet-service:[dynamodb]
// snippet-keyword:[PHP]
// snippet-keyword:[Amazon DynamoDB]
// snippet-keyword:[Code Sample]
// snippet-keyword:[ ]
// snippet-sourcetype:[full-example]
// snippet-sourcedate:[ ]
// snippet-sourceauthor:[AWS]
// snippet-start:[dynamodb.php.codeexample.SampleDataLoad] 

/**
 * Copyright 2010-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * This file is licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License. A copy of
 * the License is located at
 *
 * http://aws.amazon.com/apache2.0/
 *
 * This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
*/

require 'vendor/autoload.php';

date_default_timezone_set('UTC');

use Aws\DynamoDb\Exception\DynamoDbException;

$sdk = new Aws\Sdk([
    'region'   => 'us-west-2',
    'version'  => 'latest'
]);

$dynamodb = $sdk->createDynamoDb();

# Setup some local variables for dates

$oneDayAgo = date('Y-m-d H:i:s', strtotime('-1 days'));
$sevenDaysAgo = date('Y-m-d H:i:s', strtotime('-7 days'));
$fourteenDaysAgo = date('Y-m-d H:i:s', strtotime('-14 days'));
$twentyOneDaysAgo = date('Y-m-d H:i:s', strtotime('-21 days'));

$tableName = 'ProductCatalog';
echo "Adding data to the $tableName table...\n";

try {
    $response = $dynamodb->batchWriteItem([
        'RequestItems' => [
            $tableName => [
                [
                    'PutRequest' => [
                        'Item' => [
                            'Id'              => ['N' => '1101'],
                            'Title'           => ['S' => 'Book 101 Title'],
                            'ISBN'            => ['S' => '111-1111111111'],
                            'Authors'         => ['SS' => ['Author1']],
                            'Price'           => ['N' => '2'],
                            'Dimensions'      => ['S' => '8.5 x 11.0 x 0.5'],        
                            'PageCount'       => ['N' => '500'],        
                            'InPublication'   => ['N' => '1'],        
                            'ProductCategory' => ['S' => 'Book']
                        ]
                    ],
                ],
                [
                    'PutRequest' => [
                        'Item' => [
                            'Id'              => ['N' => '102'],
                            'Title'           => ['S' => 'Book 102 Title'],
                            'ISBN'            => ['S' => '222-2222222222'], 
                            'Authors'         => ['SS' => ['Author1', 'Author2']],
                            'Price'           => ['N' => '20'], 
                            'Dimensions'      => ['S' => '8.5 x 11.0 x 0.8'], 
                            'PageCount'       => ['N' => '600'], 
                            'InPublication'   => ['N' => '1'], 
                            'ProductCategory' => ['S' => 'Book']                    
                        ]
                    ],
                ],
                [
                    'PutRequest' => [
                        'Item' => [
                            'Id'              => ['N' => '103'], 
                            'Title'           => ['S' => 'Book 103 Title'], 
                            'ISBN'            => ['S' => '333-3333333333'], 
                            'Authors'         => ['SS' => ['Author1', 'Author2']],
                            'Price'           => ['N' => '2000'], 
                            'Dimensions'      => ['S' => '8.5 x 11.0 x 1.5'], 
                            'PageCount'       => ['N' => '600'], 
                            'InPublication'   => ['N' => '0'], 
                            'ProductCategory' => ['S' => 'Book']                  
                        ]
                    ],
                ],
                [
                    'PutRequest' => [
                        'Item' => [
                            'Id'              => ['N' => '201'], 
                            'Title'           => ['S' => '18-Bike-201'], 
                            'Description'     => ['S' => '201 Description'], 
                            'BicycleType'     => ['S' => 'Road'], 
                            'Brand'           => ['S' => 'Mountain A'], 
                            'Price'           => ['N' => '100'], 
                            'Color'           => ['SS' => ['Red', 'Black']], 
                            'ProductCategory' => ['S' => 'Bicycle']            
                        ]
                    ],
                ],
                [
                    'PutRequest' => [
                        'Item' => [
                            'Id'              => ['N' => '202'], 
                            'Title'           => ['S' => '21-Bike-202'], 
                            'Description'     => ['S' => '202 Description'], 
                            'BicycleType'     => ['S' => 'Road'], 
                            'Brand'           => ['S' => 'Brand-Company A'], 
                            'Price'           => ['N' => '200'], 
                            'Color'           => ['SS' => ['Green', 'Black']],
                            'ProductCategory' => ['S' => 'Bicycle']
                        ]
                    ],
                ],
                [
                    'PutRequest' => [
                        'Item' => [
                            'Id'              => ['N' => '203'],  
                            'Title'           => ['S' => '19-Bike-203'], 
                            'Description'     => ['S' => '203 Description'], 
                            'BicycleType'     => ['S' => 'Road'], 
                            'Brand'           => ['S' => 'Brand-Company B'], 
                            'Price'           => ['N' => '300'], 
                            'Color'           => ['SS' => ['Red', 'Green', 'Black']], 
                            'ProductCategory' => ['S' => 'Bicycle']                    
                        ]
                    ],
                ],
                [
                    'PutRequest' => [
                        'Item' => [
                            'Id'              => ['N' => '204'],  
                            'Title'           => ['S' => '18-Bike-204'], 
                            'Description'     => ['S' => '204 Description'], 
                            'BicycleType'     => ['S' => 'Mountain'], 
                            'Brand'           => ['S' => 'Brand-Company B'], 
                            'Price'           => ['N' => '400'], 
                            'Color'           => ['SS' => ['Red']], 
                            'ProductCategory' => ['S' => 'Bicycle']
                        ]
                    ],
                ],
                [
                    'PutRequest' => [
                        'Item' => [
                            'Id'              => ['N' => '205'], 
                            'Title'           => ['S' => '20-Bike-205'],
                            'Description'     => ['S' => '205 Description'],
                            'BicycleType'     => ['S' => 'Hybrid'],
                            'Brand'           => ['S' => 'Brand-Company C'],
                            'Price'           => ['N' => '500'],
                            'Color'           => ['SS' => ['Red', 'Black']],
                            'ProductCategory' => ['S' => 'Bicycle']            
                        ]
                    ]
                ]
            ],
        ],
    ]);
    echo "done.\n";
} catch (DynamoDbException $e) {
    echo $e->getMessage() . "\n";
    exit ("Unable to load data into $tableName\n");
}




$tableName = 'Forum';
echo "Adding data to the $tableName table...\n";

try {
    $response = $dynamodb->batchWriteItem([
        'RequestItems' => [
            $tableName => [
                [
                    'PutRequest' => [
                        'Item' => [
                            'Name'     => ['S' => 'Amazon DynamoDB'],
                            'Category' => ['S' => 'Amazon Web Services'],
                            'Threads'  => ['N' => '0'],
                            'Messages' => ['N' => '0'],
                            'Views'    => ['N' => '1000']
                        ]
                    ]
                ],
                [
                    'PutRequest' => [
                        'Item' => [
                            'Name'     => ['S' => 'Amazon S3'],
                            'Category' => ['S' => 'Amazon Web Services'],
                            'Threads'  => ['N' => '0']
                        ]
                    ]
                ],
            ]
        ]
    ]);
    echo "done.\n";
    
} catch (DynamoDbException $e) {
    echo $e->getMessage() . "\n";
    exit ("Unable to load data into $tableName\n");
}

$tableName = 'Thread';
echo "Adding data to the $tableName table...\n";

try {
    $response = $dynamodb->batchWriteItem([
        'RequestItems' => [
            $tableName => [
                [
                    'PutRequest' => [
                        'Item' => [
                            'ForumName'          => ['S'=>'Amazon DynamoDB'],
                            'Subject'            => ['S'=> 'DynamoDB Thread 1'],
                            'Message'            => ['S'=>'DynamoDB thread 1 message'],
                            'LastPostedBy'       => ['S'=>'User A'],
                            'LastPostedDateTime' => ['S'=>$fourteenDaysAgo],
                            'Views'              => ['N'=>'0'],
                            'Replies'            => ['N'=>'0'],
                            'Answered'           => ['N'=>'0'],
                            'Tags'               => ['SS' => ['index', 'primarykey', 'table']]
                        ]
                    ],
                    'PutRequest' => [
                        'Item' => [
                            'ForumName'          => ['S'=>'Amazon DynamoDB'],
                            'Subject'            => ['S'=> 'DynamoDB Thread 2'],
                            'Message'            => ['S'=>'DynamoDB thread 2 message'],
                            'LastPostedBy'       => ['S'=>'User A'],
                            'LastPostedDateTime' => ['S'=>$twentyOneDaysAgo],
                            'Views'              => ['N'=>'0'],
                            'Replies'            => ['N'=>'0'],
                            'Answered'           => ['N'=>'0'],
                            'Tags'               => ['SS' => ['index', 'partitionkey', 'sortkey']]
                        ]
                    ],
                    'PutRequest' => [
                        'Item' => [
                            'ForumName'          => ['S'=>'Amazon S3'],
                            'Subject'            => ['S'=> 'S3 Thread 1'],
                            'Message'            => ['S'=>'S3 Thread 3 message'],
                            'LastPostedBy'       => ['S'=>'User A'],
                            'LastPostedDateTime' => ['S'=>$sevenDaysAgo],
                            'Views'              => ['N'=>'0'],
                            'Replies'            => ['N'=>'0'],
                            'Answered'           => ['N'=>'0'],
                            'Tags'               => ['SS' => ['largeobjects', 'multipart upload']]
                        ]
                    ]
                ]
            ]
        ]
    ]);
    echo "done.\n";

} catch (DynamoDbException $e) {
    echo $e->getMessage() . "\n";
    exit ("Unable to load data into $tableName\n");
}

$tableName = 'Reply';
echo "Adding data to the $tableName table...\n";

try {
    $response = $dynamodb->batchWriteItem([
        'RequestItems' => [
            $tableName => [
                [
                    'PutRequest' => [
                        'Item' => [
                            'Id'            => ['S' => 'Amazon DynamoDB#DynamoDB Thread 1'],
                            'ReplyDateTime' => ['S' => $fourteenDaysAgo], 
                            'Message'       => ['S' => 'DynamoDB Thread 1 Reply 2 text'],
                            'PostedBy'      => ['S' => 'User B']
                        ]
                    ]
                ],
                [
                    'PutRequest' => [
                        'Item' => [
                            'Id'            => ['S' => 'Amazon DynamoDB#DynamoDB Thread 2'], 
                            'ReplyDateTime' => ['S' => $twentyOneDaysAgo], 
                            'Message'       => ['S' => 'DynamoDB Thread 2 Reply 3 text'],
                            'PostedBy'      => ['S' => 'User B']
                        ]
                    ]
                ],
                [
                    'PutRequest' => [
                        'Item' => [
                            'Id'            => ['S' => 'Amazon DynamoDB#DynamoDB Thread 2'],
                            'ReplyDateTime' => ['S' => $sevenDaysAgo],
                            'Message'       => ['S' => 'DynamoDB Thread 2 Reply 2 text'],
                            'PostedBy'      => ['S' => 'User A']
                        ]
                    ]
                ],
                [
                    'PutRequest' => [
                        'Item' => [
                            'Id'            => ['S' => 'Amazon DynamoDB#DynamoDB Thread 2'],
                            'ReplyDateTime' => ['S' => $oneDayAgo], 
                            'Message'       => ['S' => 'DynamoDB Thread 2 Reply 1 text'],
                            'PostedBy'      => ['S' => 'User A']
                        ]
                    ]
                ]
            ],
        ]
      ]);

echo "done.\n";
} catch (DynamoDbException $e) {
    echo $e->getMessage() . "\n";
    exit ("Unable to load data into $tableName\n");
}


// snippet-end:[dynamodb.php.codeexample.SampleDataLoad] 
?>